silences-gitops/README.md

```diff
@@ -0,0 +1,114 @@
+---
+
+# Deploy observability silences using GitOps
+
+The Observability Platform makes use of [Silences](https://github.com/giantswarm/silence-operator/#how-does-it-work) CRs in order to mute alerts, see [silences management](https://intranet.giantswarm.io/docs/support-and-ops/processes/silence-management/).
+
+Those Silences CRs are currently stored in a private Github repository at [https://github.com/giantswarm/silences][1], where GiantSwarm employees can create and manage silences. They are deployed to management clusters via a CronJob which pulls Silences CRs from the Github repository.
+
+## Problem statement
+
+The current deployment solution using a CronJob suffers from both privacy and security issues and does not allow customers to manage their silences. It is also not in line with any of the deployment mechanisms we currently have in place.
+
+## Solution proposal
+
+This solution proposes to re-use the current [management-cluster-base][2] (MCB) and customer management cluster (CMC) (e.g. [giantswarm-management-clusters](https://github.com/giantswarm/giantswarm-management-clusters)) GitOps mechanism in place in order to manage and deploy Silences CRs accross management clusters.
+
+This solution is laid out as follows:
+
+- Silence CRs common to all clusters are stored in the MCB repository, under the `bases/silences/` directory
+- Silence CRs added in MCB also need to be referenced in the `bases/silences/kustomization.yaml` file
+- Silence CRs specific to a cluster are stored in the corresponding CMC repository, under the `management-clusters/<management-cluster-name>/silences/` directory
+- Name collision between MCB and CMC Silences CRs is prevented by adding the special `global-` name prefix when deploying Silences CRs from the MCB repository
+
+![Silences GitOps Architecture](silences-gitops.png)
+
+### Examples
+
+- Common Silence CR in MCB repository:
+
+```yaml
+$ cat management-cluster-bases/bases/silences/alice.yaml
+apiVersion: monitoring.giantswarm.io/v1alpha1
+kind: Silence
+metadata:
+  name: alice
+  annotations:
+    valid-until: "2100-01-01"
+spec:
+  matchers:
+  - name: alertname
+    value: AliceAlert
+    isRegex: false
+```
+
+```yaml
+$ cat management-cluster-bases/bases/silences/kustomization.yaml
+resources:
+- alice.yaml
+```
+
+- Cluster specific Silence CR in CMC repository:
+
+```yaml
+$ cat giantswarm-management-clusters/management-clusters/some-management-cluster/silences/bob.yaml
+apiVersion: monitoring.giantswarm.io/v1alpha1
+kind: Silence
+metadata:
+  name: bob
+  annotations:
+    valid-until: "2100-01-01"
+spec:
+  matchers:
+  - name: alertname
+    value: BobAlert
+    isRegex: false
+  - name: cluster_id
+    value: some-cluster-id
+    isRegex: false
+```
+
+### Use cases
+
+- Create a silence for all clusters
+  - Add the Silence CR to MCB repository in `bases/silences/` directory
+  - Reference the new Silence CR in the `bases/silences/kustomization.yaml` file
```

Comment:
I think we could do without this if se use a Kustomization CR right?

Reply:
We should be able to apply the whole `bases/silences` directory

Reply:
I would prefer not to have this kustomization.yaml file, but I haven't found a workaround this.

Only the [`Kustomization.kustomize.toolkit.fluxcd.io`](https://fluxcd.io/flux/components/kustomize/api/v1/) is capable to apply a directory by generating a `kustomization.yaml` for all the yaml resources found in a directory. But this resource works by looking up the specified path in a `sourceRef` (typically a GitRepo in our case), but we have no `GitRepo` for MCB, so we can't use this method.

The other [`Kustomization.kustomize.config.k8s.io`](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/) use a `resources` field which reference location for other `kustomization.yaml` files, they can be chained but the last one must explicitly list in resources files.

Reply:
Maybe someone from @giantswarm/team-honeybadger have an idea?

Reply:
You need the `kustomization.yaml` in MCB and then reference that in CMC repos like: https://github.com/giantswarm/giantswarm-management-clusters/blob/bd8f624cb83574cb0467a5410ef0beb75d378f82/management-clusters/golem/catalogs/kustomization.yaml#L11. I think with remote bases there is no workaround that.

I am not a huge fan of the auto generate `kustomization.yaml`. That is a flux only feature and takes away rendering the manifests locally by running `kustomize`. The whole point of `kustomize` is to declare clearly what you want to apply. Flux merely runs the rendering and applies the result.

If sounds reasonable, I would also prefer a separate Flux Kustomization for silences as well. The `flux` Kustomization is getting a bit bloated. But up to you.

Reply:
> What do you mean with a separate Flux Kustomization for silences ? Here is the concept I had so far
https://github.com/giantswarm/management-cluster-bases/pull/177/files
https://github.com/giantswarm/giantswarm-management-clusters/pull/932/files

This is good, mostly, but that `silences` Flux kustomization should reconcile a `kustomziation.yaml` in the CMC repo. For that you need the replace for the path. Look at how `catalogs` or `flux-extras` Flux Kustomization work for example.

Reply:
Alright I'll look into this and I'll ask your review in the implementation.